src/connector/src/sink/snowflake.rs

```diff
@@ -177,22 +173,25 @@ pub struct SnowflakeSinkWriter {
     /// the client to insert file to external storage (i.e., s3)
     s3_client: SnowflakeS3Client,
     row_encoder: JsonEncoder,
-    row_counter: u32,
-    payload: String,
-    /// the threshold for sinking to s3
-    max_batch_row_num: u32,
     /// The current epoch, used in naming the sink files
     /// mainly used for debugging purpose
     epoch: u64,
+    /// streaming uploader, i.e., opendal s3 engine
+    streaming_uploader: Option<Box<dyn StreamingUploader>>,
```

Comment:
We can store as `Option<(StreamingUploader, String)>` to indicate that whenever there is a streaming uploader, there must be a corresponding path.

Reply:
path is internally stored in a proper-initialized `streaming-uploader`, the real purpose for separately storing `file_suffix` here is for http request to `snowflake pipe`, refer: https://github.com/risingwavelabs/risingwave/pull/16269#discussion_r1567725309.

Reply:
We can return the file name in the `finish_streaming_upload` method, so that in the later `send_request` we don't need to get the `file_suffix` from `self` with `unwrap`.

We need to carefully ensure that the `unwrap` will not cause panic. Though the current code is correct, it is error-prune like @fuyufjh said, because the correctness of the `unwrap` depends on the correct usage in a more outer scope.

Reply:
> We can store as Option<(StreamingUploader, String)> to indicate that whenever there is a streaming uploader, there must be a corresponding path.

A typical solution for this is to make a `enum State` as:

```rs
enum State {
   Init,
   StageA(... fields of state A ...),
   StageB(... fields of state B ...),
   Done,
}
```

Reply:
both approaches sound good to me - since the current state of `streaming_uploader` is relatively straightforward to manage, I prefer `Option<(StreamingUploader, String)>` + `let-else` pattern match to gracefully handle error and prevent explicit `unwrap` call(s).

if future state management of `streaming_upload` becomes complicated, we can switch to `enum State { ... }` approach to manage the state transfer.